cub/test/catch2_test_device_reduce_deterministic.cu

```diff
@@ -0,0 +1,216 @@
+    d_temp_storage,
+    temp_storage_bytes,
+    d_in,
+    d_out,
+    static_cast<OffsetT>(num_items),
+    InitT{}, // zero-initialize
+    stream);
+}
+
+template <int NOMINAL_BLOCK_THREADS_4B, int NOMINAL_ITEMS_PER_THREAD_4B>
+struct AgentReducePolicy
+{
+  /// Number of items per vectorized load
+  static constexpr int VECTOR_LOAD_LENGTH = 4;
+
+  /// Cooperative block-wide reduction algorithm to use
+  static constexpr cub::BlockReduceAlgorithm BLOCK_ALGORITHM = cub::BlockReduceAlgorithm::BLOCK_REDUCE_RAKING;
+
+  /// Cache load modifier for reading input elements
+  static constexpr cub::CacheLoadModifier LOAD_MODIFIER = cub::CacheLoadModifier::LOAD_DEFAULT;
+  constexpr static int ITEMS_PER_THREAD                 = NOMINAL_ITEMS_PER_THREAD_4B;
+  constexpr static int BLOCK_THREADS                    = NOMINAL_BLOCK_THREADS_4B;
+};
+
+template <int ItemsPerThread, int BlockSize>
+struct hub_t
+{
+  struct Policy : cub::ChainedPolicy<300, Policy, Policy>
+  {
+    constexpr static int ITEMS_PER_THREAD = ItemsPerThread;
+
+    using DeterministicReducePolicy = AgentReducePolicy<BlockSize, ItemsPerThread>;
+
+    // SingleTilePolicy
+    using SingleTilePolicy = DeterministicReducePolicy;
+
+    // SegmentedReducePolicy
+    using SegmentedReducePolicy = DeterministicReducePolicy;
+  };
+
+  using MaxPolicy = Policy;
+};
+
+DECLARE_LAUNCH_WRAPPER(DeterministicSum, deterministic_sum);
+
+// TODO (srinivasyadav18): Replace with macro `REQUIRE_APPROX_EQ_EPSILON` once the PR
+// https://github.com/NVIDIA/cccl/pull/4842 is merged
+template <typename T>
+bool approx_eq(const T& expected, const T& actual, const double tolerance = 0.01)
+{
+  double diff     = std::abs(static_cast<double>(expected) - static_cast<double>(actual));
+  double rel_diff = diff / std::abs(static_cast<double>(expected));
+  return rel_diff < tolerance;
+}
+
+C2H_TEST("Deterministic Device reduce works with float and double on gpu", "[reduce][deterministic]", float_type_list)
+{
+  using type          = typename c2h::get<0, TestType>;
+  const int num_items = 1 << 20;
+  c2h::device_vector<type> d_input(num_items);
+  c2h::gen(C2H_SEED(2), d_input, static_cast<type>(0.0), static_cast<type>(100.0));
+
+  c2h::device_vector<type> d_output(1);
+
+  const type* d_input_ptr = thrust::raw_pointer_cast(d_input.data());
+
+  deterministic_sum(d_input_ptr, d_output.begin(), num_items);
+
+  c2h::host_vector<type> h_input = d_input;
+
+  // Requires `std::accumulate` to produce deterministic result which is required for comparison
+  // with the device RFA result.
+  // NOTE: `std::reduce` is not equivalent
+  const type h_expected           = std::accumulate(h_input.begin(), h_input.end(), type{}, ::cuda::std::plus<type>());
+  c2h::host_vector<type> h_output = d_output;
+
+  REQUIRE(approx_eq(h_expected, h_output[0]));
+}
+
+template <typename type>
+void deterministic_reduce_gpu(const int N)
```

Comment:
question: why do we have to separate this into a function as opposed to just having this in the test case?